src/classes/Mfa.php

```diff
@@ -0,0 +1,178 @@
+use RobThree\Auth\TwoFactorAuth;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Session;
+use function time;
+
+/**
+ * Provide methods for multi/two-factor authentication
+ */
+class Mfa
+{
+    /** @var Session $Session Current session */
+    public $Session;
+
+    /** @var Db $Db SQL Database */
+    protected $Db;
+
+    /** @var TwoFactorAuth $TwoFactorAuth PHP Class for handling two/multi-factor authentication */
+    private $TwoFactorAuth;
+
+    /** @var Request $Request Current request */
+    private $Request;
+
+    /**
+     * Constructor
+     *
+     * @param Request $request
+     * @param Session $session
+     */
+    public function __construct(Request $request, Session $session)
+    {
+        $this->TwoFactorAuth = new TwoFactorAuth('eLabFTW', 6, 30, 'sha1', new MpdfQrProvider());
+        $this->Db = Db::getConnection();
+        $this->Request = $request;
+        $this->Session = $session;
+    }
+
+    /**
+     * Test if user has 2FA activated
+     * Redirect to multi-factor code submission if active
+     *
+     * @param string $redirect Where user will be redirected to after code submission
+     * @return void
+     */
+    public function needVerification(int $userid, string $redirect): void
+    {
+        $MFASecret = $this->getSecret($userid);
+        if ($MFASecret && !$this->Session->has('mfa_verified')) {
+            $this->Session->set('mfa_secret', $MFASecret);
+            $this->Session->set('mfa_redirect', $redirect);
+
+            $Response = new RedirectResponse('../../mfa.php');
+            $Response->send();
+            exit();
+        }
+    }
+
+    /**
+     * Load MFA secret of user from databse if exists
+     *
+     * @param int $userid
+     * @return mixed MFA secret or false
+     */
+    public function getSecret(int $userid)
+    {
+        $sql = 'SELECT mfa_secret FROM users WHERE userid = :userid';
+        $req = $this->Db->prepare($sql);
+        $req->bindParam(':userid', $userid);
+        $this->Db->execute($req);
+        $res = $req->fetchColumn();
+
+        if ($res !== null || $res !== false) {
+            return (string) $res;
+        }
+        return false;
+    }
+
+    /**
+     * Generate a new MFA secret
+     * Redirect to multi-factor code submission
+     *
+     * @param string $redirect Where user will be redirected to after code submission
+     * @return void
+     */
+    public function enable(string $redirect): void
+    {
+        // Need to request verification code to confirm user got secret and can authenticate in the future by MFA
+        $this->Session->set('mfa_secret', $this->TwoFactorAuth->createSecret(160));
+        $this->Session->set('enable_mfa', true);
+        $this->Session->set('mfa_redirect', $redirect);
+
+        $Response = new RedirectResponse('../../mfa.php');
+        $Response->send();
+        exit;
```

Comment:
there should never be an exit in a class method, return the Response and let client code deal with it.

Reply:
exit is removed